### PR TITLE
How-to: Choose Node Class

### PR DESCRIPTION
**Pull request by bastjan**
Documents how to choose node class if rolled out in [APPU-3949](https://ticket.vshn.net/browse/APPU-3949).

## Checklist

- [x] Try to isolate changes into separate PRs (to build a better changelog).
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `how-to`, `tutorial`, `reference`, `explanation`, `cicd`, `dependency`
      as they show up in the changelog

### CODE REVIEW

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 1]
```diff
@@ -0,0 +1,68 @@
+= Choose Node Class
+
+Some zones of {product} provide two or more node classes you can choose from to schedule your workload.
+For an explanation on what node group are and which one are available, please have a look at xref:references/node-classes.adoc[].
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc]:
```suggestion
For an explanation on what node groups are and which ones are available, please have a look at xref:references/node-classes.adoc[].
```

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+This page will explain to you how you can have your workload scheduled on a specific node class.
+You can do so for all workload within a given namespace.
+But you can also do so for a single pod.
+It's even possible, to a node class as default for a namespace, while selected pods run on another node class.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc]:
```suggestion
It's even possible, to add a node class as default for a namespace, while selected pods run on another node class.
```

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+You can set the node class on a namespace level by setting the `appuio.io/default-node-selector` annotation on the namespace.
+The value of the annotation is a comma separated list of key value pairs that's expanded to a node selector.
+
+[source,yaml]
+----
+apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    appuio.io/default-node-selector: "appuio.io/node-class=plus" <1>
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc:33]: Wouldn't it be nicer to be:

```
appuio.io/default-node-selector: plus
```

And then in the background expand that to `appuio.io/node-class=plus`. IMHO it's too verbose right now, and therefore prone to typos leading into support requests and potential outages.

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  name: my-namespace
+----
+<1> The node class to use for workloads in this namespace.
+
+== Setting Workload Node Selector
+
+Single Pods and their controllers can override the node class by setting the `.spec.nodeSelector` field.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc:40]: Either write "Pod" in the whole document with uppercase, or in the whole document full lowercase, but please don't mix.

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+Setting a node selector here overrides the namespace default and the cluster default.
```
Comment (tobru) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc:42]: Do we somewhere describe the "cluster default"?

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 6]
```diff
@@ -0,0 +1,32 @@
+= Node Classes
+
+Some zones of {product} provide a set of two more node classes.
+A node class is a group of nodes having a distinct set of specidifications.
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc]:
```suggestion
A node class is a group of nodes having a distinct set of specifications.
```

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc [hunk 7]
```diff
@@ -0,0 +1,67 @@
+= Choose Node Class
+
+Some zones of {product} provide two or more node classes you can choose from to schedule your workload.
+For an explanation of what node groups are and which ones are available, please have a look at xref:references/node-classes.adoc[].
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/how-to/choose-node-class.adoc:4]:
```suggestion
For an explanation of what node classes are and which ones are available, please have a look at xref:references/node-classes.adoc[].
```

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+The node classes available on a zone are indicated on the https://portal.appuio.cloud/zones[zone listing].
+
+See xref:how-to/choose-node-class.adoc[] on how to schedule your workload on a specific node class.
+
+You will find the characteristics of all node class explained below.
+
+
+flex::
++
+Flex nodes directly map to https://www.cloudscale.ch/en/pricing#flavors-explained[flex instances defined by the infrastrucutre provider].
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc:16]:
```suggestion
Flex nodes directly map to https://www.cloudscale.ch/en/pricing#flavors-explained[flex instances defined by the infrastructure provider].
```

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+They use CPUs shared with other instances.
+This allows for short term load peaks but can occasionally require your workload to wait for CPU time.
+Workload running on flex nodes will cost you less (see https://products.docs.vshn.ch/products/appuio/cloud/pricing_2022-12-01.html[pricing]).
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc:19]:
```suggestion
Workload running on Flex nodes will cost you less (see https://products.docs.vshn.ch/products/appuio/cloud/pricing.html[pricing]).
```

That points to the current one.

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
++
+Use this node class for your background jobs, test environments or applications not sensitive to latency.
++
+On an {product} zone running on cloudscale.ch, your workload will by default be scheduled on flex nodes.
+
+plus::
++
+Flex nodes directly map to https://www.cloudscale.ch/en/pricing#flavors-explained[plus instances defined by the infrastrucutre provider].
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc]:
```suggestion
Plus nodes directly map to https://www.cloudscale.ch/en/pricing#flavors-explained[plus instances defined by the infrastructure provider].
```

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+Plus nodes have dedicated CPU cores.
+This gives you consistent performance but will cost you more (see https://products.docs.vshn.ch/products/appuio/cloud/pricing_2022-12-01.html[pricing]).
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc:29]:
```suggestion
This gives you consistent performance but will cost you more (see https://products.docs.vshn.ch/products/appuio/cloud/pricing.html[pricing]).
```

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
++
+The plus node class is probably the right choise for your production frontends, database and other applications sensitive to latency.
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc]:
```suggestion
The plus node class is probably the right choice for your production frontends, database and other applications sensitive to latency.
```

File: docs/modules/ROOT/pages/references/node-classes.adoc [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+Some zones of {product} provide a set of two more node classes.
+A node class is a group of nodes having a distinct set of specidifications.
+Those nodes can be optimized for compute, memory or cost or they could offer special features like GPU support.
+
+The node classes available on a zone are indicated on the https://portal.appuio.cloud/zones[zone listing].
+
+See xref:how-to/choose-node-class.adoc[] on how to schedule your workload on a specific node class.
+
+You will find the characteristics of all node class explained below.
+
```
Comment (tobru) [inline docs/modules/ROOT/pages/references/node-classes.adoc:12]: I'd add a heading here:

```suggestion
== cloudscale.ch Zones
```

**Review by tobru — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc (hunk 3 above)
Comment (corvus-ch) [reply docs/modules/ROOT/pages/how-to/choose-node-class.adoc:33]: If I understand correctly, this is mimicking how node selectors work. @bastjan might be able to elaborate.

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc (hunk 5 above)
Comment (tobru) [reply docs/modules/ROOT/pages/how-to/choose-node-class.adoc:42]: That would be a good idea :+1:

File: docs/modules/ROOT/pages/how-to/choose-node-class.adoc (hunk 3 above)
Comment (tobru) [reply docs/modules/ROOT/pages/how-to/choose-node-class.adoc:33]: It's fine by me to not open a can of worms here, it's probably just fine. Let's get this feature out :+1:

**Review by tobru — APPROVED** [review]

**Pull request merged**